Fixes race condition on ELB checking instance availability (reported in #383)

jsmartin opened the pull request:
This patch uses the ELB as the source of truth (when applicable) for the health state of an instance in an ASG.  This negates any race condition.

This will also wait for a new ASG to have it's instances be healthy, as deemed by the ELB.

@bwhaley please review.

bwhaley commented:
Would like to review but am unable until 1/17 or so due to vacation.

Sent from my iPhone

> On Jan 6, 2015, at 1:24 PM, James S. Martin notifications@github.com wrote:
> 
> This patch uses the ELB as the source of truth (when applicable) for the health state of an instance in an ASG. This negates any race condition.
> 
> This will also wait for a new ASG to have it's instances be healthy, as deemed by the ELB>
> 
> @bwhaley please review.
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/jsmartin/ansible-modules-core asg_rolling_check_lb
> Or view, comment on, or merge it at:
> 
>   https://github.com/ansible/ansible-modules-core/pull/601
> 
> Commit Summary
> 
> vpc_zone_identifier must be a csv string when an asg is updated.
> check lb for instance healthy and some refactoring of rolling replace functions.
> File Changes
> 
> M cloud/amazon/ec2_asg.py (94)
> Patch Links:
> 
> https://github.com/ansible/ansible-modules-core/pull/601.patch
> https://github.com/ansible/ansible-modules-core/pull/601.diff
> —
> Reply to this email directly or view it on GitHub.

jsmartin commented:
no problem. enjoy the vacay.

Commit:
check lb for instance healthy and some refactoring of rolling replace functions.

Optionally waiting for new instances to be in viable state on ASG creation/update.

Properly updating properties and terminating group.

bwhaley commented:
Per discussion via IRC, @jsmartin will add a call to `wait_for_elb()` for new ASGs in addition to when replacing instances. The logic looks good otherwise.

jsmartin commented:
@bwhaley Added some new commits and docs.  wait_for_elb now optionally happens on new/updated ASGs as well.

bwhaley commented:
I haven't had a chance to test but code LGTM!

rwky commented:
This doesn't work with replacing instances, I've submitted a pull request to @jsmartin repo jsmartin/ansible-modules-core/pull/1

abadger commented on cloud/amazon/ec2_asg.py:
```diff
@@ -403,6 +461,8 @@ def delete_autoscaling_group(connection, module):
 
         group.delete()
+        while len(connection.get_all_groups(names=[group_name])):
+                time.sleep(5)
```
Extra indent here.

abadger commented:
Okay, I rebased this onto the tip of devel and then merged this in.  @rwky, I also pulled in your changes but I squashed them into a single commit.  The rebasing and squashing were because I also pulled this into the stable-1.9 branch and it was easier to cherry-pick the commits that way.  thanks guys for creating the code and for testing and reviewing each others' changes.

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
